File: content/snippets/scala3/interview_lcci_08_14_boolean_evaluation.scala
object Lcci0814BooleanEvaluation:
    val MOD = 1000000007L

    def countEval(s: String, result: Int): Int =
        val n = s.length
        // dp(i)(j)(v) = ways s[i..j] evaluates to v
        val dp = Array.ofDim[Long](n, n, 2)
        for (i <- 0 until n by 2) dp(i)(i)(s(i) - '0') = 1L
        var size = 3
        while (size <= n) {
            var i = 0
            while (i <= n - size) {
                val j = i + size - 1
                var k = i + 1
                while (k < j) {
                    val op = s(k)
                    val lf = dp(i)(k-1)(0); val lt = dp(i)(k-1)(1)
                    val rf = dp(k+1)(j)(0); val rt = dp(k+1)(j)(1)
                    op match {
                        case '&' =>
                            dp(i)(j)(1) = (dp(i)(j)(1) + lt * rt) % MOD
                            dp(i)(j)(0) = (dp(i)(j)(0) + lf*rf + lf*rt + lt*rf) % MOD
                        case '|' =>
                            dp(i)(j)(1) = (dp(i)(j)(1) + lt*rt + lf*rt + lt*rf) % MOD
                            dp(i)(j)(0) = (dp(i)(j)(0) + lf*rf) % MOD
                        case '^' =>
                            dp(i)(j)(1) = (dp(i)(j)(1) + lf*rt + lt*rf) % MOD
                            dp(i)(j)(0) = (dp(i)(j)(0) + lf*rf + lt*rt) % MOD
                    }
                    k += 2
                }
                i += 2
            }
            size += 2
        }
        dp(0)(n-1)(result).toInt

    @main def main(): Unit =
        println(countEval("1^0|0|1", 0))      // 2
        println(countEval("0&0&0&1^1|0", 1))  // 10

File: content/snippets/typescript/interview_lcci_08_14_boolean_evaluation.ts
const MOD = BigInt(1_000_000_007);

function countEval(s: string, result: number): number {
    const n = s.length;
    // dp[i][j] = [ways_false, ways_true]
    const dp: [bigint, bigint][][] = Array.from({length: n}, () =>
        Array.from({length: n}, () => [0n, 0n] as [bigint, bigint])
    );
    for (let i = 0; i < n; i += 2) {
        dp[i][i][s[i] === '1' ? 1 : 0] = 1n;
    }
    for (let size = 3; size <= n; size += 2) {
        for (let i = 0; i <= n - size; i += 2) {
            const j = i + size - 1;
            for (let k = i + 1; k < j; k += 2) {
                const op = s[k];
                const lf = dp[i][k-1][0], lt = dp[i][k-1][1];
                const rf = dp[k+1][j][0], rt = dp[k+1][j][1];
                if (op === '&') {
                    dp[i][j][1] = (dp[i][j][1] + lt * rt) % MOD;
                    dp[i][j][0] = (dp[i][j][0] + lf*rf + lf*rt + lt*rf) % MOD;
                } else if (op === '|') {
                    dp[i][j][1] = (dp[i][j][1] + lt*rt + lf*rt + lt*rf) % MOD;
                    dp[i][j][0] = (dp[i][j][0] + lf*rf) % MOD;
                } else { // '^'
                    dp[i][j][1] = (dp[i][j][1] + lf*rt + lt*rf) % MOD;
                    dp[i][j][0] = (dp[i][j][0] + lf*rf + lt*rt) % MOD;
                }
            }
        }
    }
    return Number(dp[0][n-1][result]);
}

console.log(countEval("1^0|0|1", 0));      // 2
console.log(countEval("0&0&0&1^1|0", 1));  // 10
